### PR TITLE
Node base Role that runs on all nodes to

### LINKED ISSUE #2082: Let cluster controller heartbeat not only workers but all nodes
Currently, we only heartbeat nodes that run the `Worker` role. This should change now and include all nodes.

### PR DESCRIPTION
**Pull request by muhamadazmy**
Node base Role that runs on all nodes to

Summary:
Introduce a new `base` role that is hosted
by all nodes. This allows to handle common
node related messages like `GetNodeState`

Fixes #2082

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/2157).
* #2158
* __->__ #2157

### CODE REVIEW

File: crates/node/src/roles/base.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+use restate_core::{
+    cancellation_watcher,
+    network::{Incoming, MessageRouterBuilder, NetworkError},
+    task_center,
+    worker_api::ProcessorsManagerHandle,
+    ShutdownError, TaskKind,
+};
+use restate_types::net::node::{GetNodeState, NodeStateResponse};
+
+pub struct BaseRole {
```
Comment (pcholakov) [inline crates/node/src/roles/base.rs:25]: Question for all: how should we think about putting responsibilities here vs. in the gRPC `NodeSvc` service? When is one a better choice than the other?

Naming: is this really a role, or just a baseline `NodeService`?

**Review by pcholakov — COMMENTED** [review]
This makes sense to me, only raising some questions in the interest of improving my understanding - and potentially getting to even clearer naming for the long-run.

Thanks for working on improving this!

File: crates/types/src/net/node.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+use super::TargetName;
+use crate::{cluster::cluster_state::PartitionProcessorStatus, identifiers::PartitionId};
+
+super::define_rpc! {
+    @request=GetNodeState,
+    @response=NodeStateResponse,
+    @request_target=TargetName::NodeGetNodeStateRequest,
+    @response_target=TargetName::NodeGetNodeStateResponse,
+}
```
Comment (pcholakov) [inline crates/types/src/net/node.rs:24]: I'm curious about how you (and maybe @AhmedSoliman when available to comment) see this mechanism vs. the new `IdentResponse`? Do we need more than one way to communicate node status info?
 https://github.com/restatedev/restate/blob/18d3b3d7bdf1df546a65e4bf412fa7c49fc2709e/crates/core/protobuf/node_svc.proto#L31-L48

Comment (muhamadazmy) [reply crates/types/src/net/node.rs:24]: Yes, I think there is a difference between the two ways (as far as I understand of course) is that `IdentResposne` is intended for grpc use and hence can be used by external tools like `restatectl` to make call directly to the intended node.

On the other hand GetNodeState is intended for node to node communication and hence can be only called vi the node networking. It can be used for example (as per this PR) by the ClusterController to collect all nodes state to build a view on the status of all nodes. 

Basically all messages that are defined with the `defeine_rpc!` macro is used for cross node communication.

File: crates/node/src/roles/base.rs (hunk 1 above)
Comment (muhamadazmy) [reply crates/node/src/roles/base.rs:25]: based on my understanding of the node networking. the BaseRole can serve peer to peer communication across nodes, which is powered by the node core `networking`. In this model messages can be exchanged knowing only the NodeID.

A gRPC service like NodeSvc on the other hand can serve external clients requests (like restatectl) to answer questions like GetIdent, ClusterState, etc... and in this model you need to know which IP and port the service is listening

File: crates/worker/src/lib.rs [hunk 3]
```diff
@@ -21,6 +21,7 @@ mod subscription_controller;
 mod subscription_integration;
 
 use codederror::CodedError;
+use restate_core::worker_api::ProcessorsManagerHandle;
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/worker/src/lib.rs:24]: Always 🤦🏼‍♂️

File: crates/types/src/net/node.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
+pub struct GetNodeState {}
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/types/src/net/node.rs:27]: I wasn't sure either, so i used the (now deprecated) message `GetProcessorsState` as a guide

[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/types/src/net/node.rs:27]: but I am okay to rename it, I also think `Status` makes more sense

File: crates/types/src/net/node.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct NodeStateResponse {
+    /// State of paritions processor per parition. Is set to None if this node is not a `Worker` node
+    #[serde_as(as = "Option<serde_with::Seq<(_, _)>>")]
+    pub paritions_processor_state: Option<BTreeMap<PartitionId, PartitionProcessorStatus>>,
+}
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/types/src/net/node.rs:35]: Yes, that's _my_ intention. I assumed in the future we will can gather more information about all the different `roles` that are running on the node.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @muhamadazmy. The changes look good to me. The one question I had was whether we can get rid of the old rpc request/response to get the partition processor state. Apart from this, +1 for merging.

File: crates/core/src/worker_api/partition_processor_manager.rs [hunk 6]
```diff
@@ -8,16 +8,22 @@
...
 
+use std::collections::BTreeMap;
+
 use tokio::sync::{mpsc, oneshot};
 
-use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::{
+    cluster::cluster_state::PartitionProcessorStatus,
+    identifiers::{PartitionId, SnapshotId},
+};
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/core/src/worker_api/partition_processor_manager.rs:18]: That is already done in #2158

File: crates/types/src/net/node.rs (hunk 2 above)
Comment (pcholakov) [reply crates/types/src/net/node.rs:24]: A bit slow to come back to this PR - thanks for the answer, this is very helpful context I was missing!
